src/connectivity/separators.c

```diff
@@ -34,97 +34,253 @@
 
 #include "core/interruption.h"
 
-static igraph_error_t igraph_i_is_separator(const igraph_t *graph,
-                                 igraph_vit_t *vit,
-                                 igraph_integer_t except,
-                                 igraph_bool_t *res,
-                                 igraph_vector_bool_t *removed,
-                                 igraph_dqueue_int_t *Q,
-                                 igraph_vector_int_t *neis,
-                                 igraph_integer_t no_of_nodes) {
-
-    igraph_integer_t start = 0;
-
-    if (IGRAPH_VIT_SIZE(*vit) >= no_of_nodes - 1) {
-        /* Just need to check that we really have at least n-1 vertices in it */
-        igraph_vector_bool_t hit;
-        igraph_integer_t nohit = 0;
-        IGRAPH_CHECK(igraph_vector_bool_init(&hit, no_of_nodes));
-        IGRAPH_FINALLY(igraph_vector_bool_destroy, &hit);
-        for (IGRAPH_VIT_RESET(*vit);
-             !IGRAPH_VIT_END(*vit);
-             IGRAPH_VIT_NEXT(*vit)) {
-            igraph_integer_t v = IGRAPH_VIT_GET(*vit);
-            if (!VECTOR(hit)[v]) {
-                nohit++;
-                VECTOR(hit)[v] = true;
-            }
-        }
-        igraph_vector_bool_destroy(&hit);
-        IGRAPH_FINALLY_CLEAN(1);
-        if (nohit >= no_of_nodes - 1) {
-            *res = false;
-            return IGRAPH_SUCCESS;
+
+/**
+ * \function igraph_i_is_separator
+ *
+ * Checks if vertex set \c S is a separator. Optionally, also checks if it
+ * is a _minimal_ separator.
+ *
+ * \param graph The graph, edge directions are ignored.
+ * \param S Candidate vertex set.
+ * \param is_separator Pointer to boolean, is S a separator?
+ * \param is_minimal If not \c NULL, it is also checked whether the separator
+ *    is minimal. This takes additional time.  If S is not a separator, this is
+ *    set to false
+ */
+static igraph_error_t igraph_i_is_separator(
+    const igraph_t *graph,
+    igraph_vs_t S,
+    igraph_bool_t *is_separator,
+    igraph_bool_t *is_minimal
+) {
+
+    const igraph_integer_t vcount = igraph_vcount(graph);
+
+    /* bit 0: is it visited?
+     * bit 1: is it in S?
+     * bit 2: marks the reachability of vertices in S in the minimal separator test
+     */
+    igraph_vector_char_t mark;
+
+    igraph_vector_int_t neis;
+    igraph_dqueue_int_t Q;
+    igraph_vit_t vit;
+    igraph_integer_t S_size = 0, S_visited_count = 0;
+
+    IGRAPH_CHECK(igraph_vit_create(graph, S, &vit));
+    IGRAPH_FINALLY(igraph_vit_destroy, &vit);
+
+    IGRAPH_VECTOR_CHAR_INIT_FINALLY(&mark, vcount);
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&neis, 10);
+    IGRAPH_DQUEUE_INT_INIT_FINALLY(&Q, 100);
+
+#define VISITED(x) (VECTOR(mark)[x] & 1) /* Was x visited? */
+#define SET_VISITED(x) (VECTOR(mark)[x] |= 1) /* Mark x as visited. */
+#define INS(x) (VECTOR(mark)[x] & 2) /* Is x in S? */
```

Comment:
I'd call this `IN_S`; `INS` makes me think about insertion.

Reply:
Sure, I'll change that.

The minimal separator test is wrong, I'm working on fixing that now.